hilti/runtime/include/deferred-expression.h

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2020-2023 by the Zeek Project. See LICENSE for details.
+
+#pragma once
+
+#include <string>
+#include <utility>
+
+#include <hilti/rt/extension-points.h>
+#include <hilti/rt/types/result.h>
+
+namespace hilti::rt {
+
+/**
+ * Wrapper for an expression that's evaluation is deferred until requested.
+ * The expression must be wrapped into a function call, and it's evaluation
+ * is requested through the wrapper's call operator.
+ *
+ * The function should be stateless as it might be invoked an unspecified
+ * number of times.
+ */
+template<typename Result, typename Expr>
+class DeferredExpression {
+public:
+    DeferredExpression(Expr&& expr) : _expr(std::move(expr)) {}
```

Comment:
I believe `Expr&&` is an universal reference here, not necessarily an r-value.

```suggestion
    DeferredExpression(Expr&& expr) : _expr(std::forward<Expr>(expr)) {}
```

Reply:
Had that, but that actually didn't compile for me.